### ISSUE TITLE
Question about OAuth negotiation

### ISSUE THREAD

**Issue by arunranga** [carrot/go-pinterest#34]
@BrandonRomano : does this lib presume an access token is already available, or does it also handle the OAuth negotiation starting with an AppID?

**Comment by BrandonRomano** [comment]
@arunranga -- No this API does not handle OAuth negotiation.  I also don't think this library should be responsible for that, but that is up for discussion!

How you currently register the access token to the client is with this method:
https://github.com/carrot/go-pinterest/blob/master/pinterest.go#L50-L53

In use it looks like this:

```go
client := pinterest.NewClient().
        RegisterAccessToken("USERS_ACCESS_TOKEN")
```

**Comment by arunranga** [comment]
@BrandonRomano OK, so the way this would be done is:

1. Ping oauth endpoint namely api.pinterest.com/oauth
2. After negotiation, obtain token
3. Stash token for reuse with golang lib/method above. What's the practice you folks use for this? I *don't think* we should encourage client cookies or "in the clear over URL" but... 

I see it both ways. Token stashing is a hard-ish prob, so maybe having a strictly call-focused lib is good. On the other hand, it's also pretty useful for devs.

**Comment by BrandonRomano** [comment]
So, the problematic part of adding that to this library is within step `#2` - specifically `negotiation`. 

To my understanding, the only way to accomplish the negotiation part of an OAuth flow would be for this library to actually run a contactable server for Pinterests OAuth flow to respond to. This library itself does not run a server, nor do I think it should make the assumption that it will be used by something that is running a server.

If your concern is coming from the fact that there is too much friction in developers getting a token to get started, I should probably add the Pinteret token generator to the README of this project ( https://developers.pinterest.com/tools/access_token/ )!

**Comment by arunranga** [comment]
Oh, the `negotiation` portion of the lib wouldn't use an actual URL; rather, the library would be set up to handle the protocol and obtain a token. In other words, there'd be method calls based on each step of the protocol. Is that wrong thinking? It disagrees with your assertion that it shouldn't run on anything that presumes a server. It could be a secondary module. I think folks use FastCGI to integrate w GoLang through some mods_rewrite magic, so it could be useful to them. Just a suggestion, though -- this is great work.

The token generator is useful to immediately get up and running.

**Comment by BrandonRomano** [comment]
Just so I get an idea, @arunranga -- what would the user be expected to pass through to the API client in order to successfully go through the flow you suggest?  Their username/password?  

I'm curious as to what part of the process you imagine the user authenticates themselves.

**Comment by arunranga** [comment]
@BrandonRomano maybe I'm thinking about this wrong.

I was trying to see if there's anything that would do the equivalent of

`if (!params.accessToken) {
      accessToken = Auth.getAccessToken();
    }`

that could be done outside the client. Maybe there isn't an additional convenience layer, other than maybe some lightweight JS (lighter than Pinterest SDK) to simplify client-side oauth piece and go along with GoLang piece on backend, and no tokens in query strings, etc.

**Comment by BrandonRomano** [comment]
Yeah, so something like that definitely makes sense in a Javascript SDK, as that can start up the OAuth process if the Javascript is running client-side.

But this code isn't going to be running client-side, so there's no generic way we can actually attach ourselves to the browser and tell it to start the oAuth flow without making large assumptions about the users Frontend / Go setup.

I can't think of any additional convenience layer we can add to this to get the token, while still keeping this library usable for _any_ Go setup.  The _only_ way I can think of is with a websocket + running a server to communicate with a browser, but that won't even work with the simplest of setups, and makes the assumption you have a frontend, and that  frontend supports websockets.

**Comment by BrandonRomano** [comment]
@arunranga, visiting this again:

Just want to verify something, as I believe we _might_ have been talking about two different things... Looking at https://developers.pinterest.com/docs/api/overview/?:

The `Getting your authorization code` section is what I was talking about that couldn't be done with this library, as there needs to be a `redirect_uri`, which this library itself does not manage.

There is also a `Getting your access token` section, which you obtain an access token using your `authorization code` obtained from the above.

The latter I am going to implement, but it will still not allow us to pass in an `AppID` to go through the entire OAuth flow.

**Comment by arunranga** [comment]
Yep same understanding. Makes sense.

**Comment by BrandonRomano** [comment]
Okay, perfect 👌 Just about wrapped up #35, so what I described will be a feature soon.  Will close this issue when that PR is merged.

**Comment by arunranga** [comment]
@BrandonRomano, when you merge I'll pull. Dumb question, but what is a good
hosting environment to run Go as backend code? E.g. suppose I were to
deploy an environment similar to what Carrot did for Target, what server /
hosting / tech configs would I use?

**Comment by BrandonRomano** [comment]
Hey @arunranga -- Totally a legitimate question, but I'm going to move that conversation over to email (as that conversation is out of scope of this repo).

Closing this now, as it's fixed by #35
